web3.nim

```diff
@@ -655,23 +662,57 @@ macro contract*(cname: untyped, body: untyped): untyped =
 
         result.add quote do:
           type `cbident` = object
-          proc subscribe(s: Sender[`cname`], t: typedesc[`cbident`], options: JsonNode, `callbackIdent`: `procTy`): Future[Subscription] =
-            let options = addAddressAndSignatureToOptions(options, s.contractAddress, "0x" & toLowerAscii($keccak256.digest(`signature`)))
+
+          template eventTopic*(T: type `cbident`): string =
+            "0x" & toLowerAscii($keccak256.digest(`signature`))
+
+          proc subscribe(s: Sender[`cname`],
+                         t: type `cbident`,
+                         options: JsonNode,
+                         `callbackIdent`: `procTy`): Future[Subscription] =
+            let options = addAddressAndSignatureToOptions(options, s.contractAddress, eventTopic(`cbident`))
 
             s.web3.subscribeToLogs(options) do(`jsonIdent`: JsonNode):
               `argParseBody`
               `call`
 
-          proc subscribe(s: Sender[`cname`], t: typedesc[`cbident`], options: JsonNode, `callbackIdent`: `procTyWithRawData`): Future[Subscription] =
-            let options = addAddressAndSignatureToOptions(options, s.contractAddress, "0x" & toLowerAscii($keccak256.digest(`signature`)))
+          proc subscribe(s: Sender[`cname`],
+                         t: type `cbident`,
+                         options: JsonNode,
+                         `callbackIdent`: `procTyWithRawData`): Future[Subscription] =
+            let options = addAddressAndSignatureToOptions(options, s.contractAddress, eventTopic(`cbident`))
 
             s.web3.subscribeToLogs(options) do(`jsonIdent`: JsonNode):
               `argParseBody`
               `callWithRawData`
-
     else:
       discard
 
+  when defined(debugMacros) or defined(debugWeb3Macros):
+    echo result.repr
+
+proc getJsonLogs*(s: Sender,
+                  EventName: type,
+                  fromBlock, toBlock = none(RtBlockIdentifier),
```

Comment:
or just add two overloads so that it's not possible to call with the wrong parameter groups?